### ISSUE TITLE
Training stops too early and selects wrong model

### ISSUE THREAD

**Issue by stweil** [mittagessen/kraken#297]
I'd expect that stage 22 produced the best model and that there was no reason for an early stop:

```
Initializing model ✓
stage 1/∞  [####################################]  6686/6686          Accuracy report (1) 0.5508 22914 10293
stage 2/∞  [####################################]  6686/6686          Accuracy report (2) 0.6767 22914 7409
stage 3/∞  [####################################]  6686/6686          Accuracy report (3) 0.7305 22914 6175
stage 4/∞  [####################################]  6686/6686          Accuracy report (4) 0.7625 22914 5442
stage 5/∞  [####################################]  6686/6686          Accuracy report (5) 0.7833 22914 4965
stage 6/∞  [####################################]  6686/6686          Accuracy report (6) 0.7944 22914 4710
stage 7/∞  [####################################]  6686/6686          Accuracy report (7) 0.8102 22914 4349
stage 8/∞  [####################################]  6686/6686          Accuracy report (8) 0.8218 22914 4083
stage 9/∞  [####################################]  6686/6686          Accuracy report (9) 0.8271 22914 3961
stage 10/∞  [####################################]  6686/6686          Accuracy report (10) 0.8346 22914 3790
stage 11/∞  [####################################]  6686/6686          Accuracy report (11) 0.8388 22914 3694
stage 12/∞  [####################################]  6686/6686          Accuracy report (12) 0.8400 22914 3667
stage 13/∞  [####################################]  6686/6686          Accuracy report (13) 0.8425 22914 3608
stage 14/∞  [####################################]  6686/6686          Accuracy report (14) 0.8476 22914 3493
stage 15/∞  [####################################]  6686/6686          Accuracy report (15) 0.8509 22914 3416
stage 16/∞  [####################################]  6686/6686          Accuracy report (16) 0.8466 22914 3515
stage 17/∞  [####################################]  6686/6686          Accuracy report (17) 0.8566 22914 3286
stage 18/∞  [####################################]  6686/6686          Accuracy report (18) 0.8555 22914 3312
stage 19/∞  [####################################]  6686/6686          Accuracy report (19) 0.8589 22914 3233
stage 20/∞  [####################################]  6686/6686          Accuracy report (20) 0.8578 22914 3258
stage 21/∞  [####################################]  6686/6686          Accuracy report (21) 0.8593 22914 3224
stage 22/∞  [####################################]  6686/6686          Accuracy report (22) 0.8630 22914 3140

Moving best model uat_17.mlmodel (0.8565942049026489) to uat_best.mlmodel
```

**Comment by stweil** [comment]
My 2nd training finished and selected a model with 0.9282 accuracy as best one, although there were was a better one with 0.9284 accuracy.

**Comment by mittagessen** [comment]
The idea is that any improvement beyond the last significant one that reset the lag counter is probably just noise so it chooses that one instead of a later one with marginally better scores. But I'm getting tired of explaining that so I'll just change it to the absolute best one.

**Comment by bencomp** [comment]
It looks like the default `min_delta` hyperparameter is `None`, instead of 0.005 as the documentation says. Maybe that causes kraken to not reset the lag counter at all?

**Comment by mittagessen** [comment]
On 21/11/12 08:36AM, Ben Companjen wrote:
> It looks like the default `min_delta` hyperparameter is `None`,
> instead of 0.005 as the documentation says. Maybe that causes kraken
> to not reset the lag counter at all?

`None` is the value for the autoscaling of the threshold. This works as
long as the potentially achievable accuracy is fairly high (>95-97%) but
will cause premature termination if the dataset is too small or the
model is too small for the dataset.

The website docs are a bit out of date. I had updated them at some point
but must have accidentally deleted the branch without merging it. I'm
currently in the process of rewriting them again.

**Comment by mittagessen** [comment]
> I'm not sure I understand the explanation though.

The idea is that what is considered a significant improvement that would reset the counter changes based on the circumstances. If the accuracy is already at 99.9% an improvement by 0.05 percentage points is probably actually meaningful while at lower accuracy levels changes like these are usually just random fluctuations (and eventually overfitting). So the algorithm scales the threshold for resetting the counter for the patience period by the potential improvement: `(1 - best_loss) / 20`.
